File: 2019/day 02/day_02.clj
; day_02.clj

(defn run 
    ([values] 
        (run values 0)
    )
    ([values start]    
        (let [op_code (get values start)]
            (cond 
                (< op_code 99)
                    (let [
                        a      (get values (get values (+ start 1)))
                        b      (get values (get values (+ start 2)))
                        output (get values (+ start 3))
                        ] 
                        (cond
                            (= op_code 1) (recur (assoc values output (+ a b)) (+ start 4))
                            (= op_code 2) (recur (assoc values output (* a b)) (+ start 4))
                            :else (assert false)
                        )
                    )
                        
                (= op_code 99) values
                :else (assert false)
            )
        )
    )
)

(assert (= (run [1,9,10,3,2,3,11,0,99,30,40,50]) [3500 9 10 70 2 3 11 0 99 30 40 50]))
(assert (= (run [1,0,0,0,99]) [2,0,0,0,99]))
(assert (= (run [2,3,0,3,99]) [2,3,0,6,99]))
(assert (= (run [2,4,4,5,99,0]) [2,4,4,5,99,9801]))
(assert (= (run [1,1,1,4,99,5,6,0,99]) [30,1,1,4,2,5,6,0,99]))

(defn toInt [x]
    (Integer/parseInt x)
)

(def data_txt (slurp "input.txt"))
(def data_splitted (clojure.string/split (clojure.string/trim-newline data_txt) #","))
(def data (vec (map toInt data_splitted)))

(defn set_noun_verb [data noun verb]
    (assoc (assoc data 1 noun) 2 verb)
)

(defn find_solution [data noun verb]
    (get (run (set_noun_verb data noun verb)) 0)
)

(println "Part 1:" (find_solution data 12 2))

; Part 2
(loop [noun 0 verb 0]    
    (if (= (find_solution data noun verb) 19690720)
        (println "Part 2:" (+ (* 100 noun) verb))
        (if (= noun 100)
            (recur 0 (inc verb))
            (recur (inc noun) verb)
        )
    )
)

File: 2019/day 02/day 02.py
"""
Day 2 puzzle - Advent of Code 2019
Pieter Kitslaar
"""
def txt_values(txt):
    return [int(v) for v in txt.split(',')]

def run(in_values, noun_verb = None):
    values = in_values[:]
    if noun_verb:
        noun, verb = noun_verb
        values[1] = noun
        values[2] = verb
    current_pos = 0
    while True:
        op_code = values[current_pos]
        if op_code < 99:
            a, b, output = values[current_pos+1:current_pos+4]
            if op_code == 1:
                # addition
                values[output] = values[a] + values[b]
            elif op_code == 2:
                # multiply
                values[output] = values[a] * values[b]
            else:
                raise ValueError("Invalid op_code {op_code} at position {current_pos}")
        elif op_code == 99:
            break
        current_pos += 4
    return values

def test_examples():
    for in_, out_ in [
                ("1,9,10,3,2,3,11,0,99,30,40,50", "3500,9,10,70,2,3,11,0,99,30,40,50"),
                ("1,0,0,0,99", "2,0,0,0,99"),
                ("2,3,0,3,99", "2,3,0,6,99"),
                ("2,4,4,5,99,0", "2,4,4,5,99,9801"),
                ("1,1,1,4,99,5,6,0,99", "30,1,1,4,2,5,6,0,99"),
                
            ]:
        try:
            assert(txt_values(out_) == run(txt_values(in_)))
        except Exception as e:
            print(in_, out_)
            raise

def main():
    # Part 1
    with open(Path(__file__).parent / 'input.txt', 'r') as f:
        part1_txt = f.read()
    part1_data = txt_values(part1_txt)
    part1_output = run(part1_data, (12, 2))
    print(part1_output[0])

    # Part 2
    for noun in range(100):
        for verb in range(100):
            output = run(part1_data, (noun, verb))
            if output[0] == 19690720:
                print('noun', noun, 'verb', verb, 'result:', 100 * noun + verb)
                break
         